content/post/asdf.md

```diff
@@ -0,0 +1,250 @@
+are written in Go and otherwise it will be hard to integrate.
+7. On-the-fly compression. NLP models often contain strings which can require much
+space while being perfectly compressible. Integers can be compressed too since
+we sometimes don't know their range beforehand and use 32 bits while only 16 are
+really needed.
+
+[HDF5](https://support.hdfgroup.org/HDF5/) was the closest to those. It is binary,
+there is no schema, it supports big tensors, Python bindings are mature and
+well-integrated, there are bindings for other languages. HDF5 is used in e.g. [Keras](https://github.com/keras-team/keras).
+However, it is not ideal in terms of performance and lacks some modern features.
+Read ["Moving away from HDF5"](https://cyrille.rossant.net/moving-away-hdf5/).
+
+We previously had some positive experience with SQLite + SQLAlchemy on top,
+but of course that variant does not stand big data blobs (4).
+There is a common workaround for (4): it is possible to store huge tensors as external files.
+This implies concatenating all the blocks together before uploading and splitting
+them back after downloading, e.g. as a TAR or a ZIP without compression.
+Those meta-archives are typically used as a temporary transfer medium
+(e.g., [TensorFlow Hub](https://www.tensorflow.org/hub/hosting)).  
+In turn, this means the higher usage complexity, the doubled requirement for free disk size,
+the increased vulnerability to data corruptions and the explosion on the number
+of open file descriptors.
+
+No more intrigue: we discovered ASDF.
+
+## ASDF
+
+[Advanced Scientific Data Format](https://github.com/spacetelescope/asdf) (ASDF)
+is a next generation serialization format for scientific data. This means that
+it focuses on storing sparse and dense tensors in an efficient way.
+The ASDF project started by [Michael Droettboom](https://github.com/mdboom) (Matplotlib; astropy)
+at SpaceTelescope Institute in 2014. ASDF is not implementation-driven, and it is based on
+the well-defined [standard](https://asdf-standard.readthedocs.io/en/latest/).
+However, there is only one maintained software library written in Python.
+While historically ASDF targeted astronomers, it is actually abstracted away
+from any scientific domain and is completely versatile. ASDF features:
+
+* Transparent, automatic, on-the-fly compression and decompression with zlib, bzip2 or lz4.
+* YAML header with binary blocks appended to the tail. All the perks of using YAML are preserved.
+* Uncompressed tensors can be [memory mapped](https://en.wikipedia.org/wiki/Memory-mapped_file) so that the operating memory consumption is very low with tiny performance penalty for sequential read and write. A killer feature if your tensors are big.
+* Data structure can be validated with YAML schemas.
+* Python and numpy arrays are first-class citizens.
+* The tagging mechanism allows to extend for new binary data types easily. Though it is rarely needed in practice.
```

Comment:
I've been wondering about this: do you guys use custom types at all?

Reply:
We don't, but not because we don't want. We cut the corners in the beginning and now need to refactor. So we leverage advanced features gradually.